### DISCUSSION TITLE
Would it be possible to type narrow overloads after the fact?

### DISCUSSION THREAD

**Discussion by JanEricNitschke** [microsoft/pyright#8549] [Ideas]
I have a case where i have an overloaded function where the different return types largely overlap. So regardless of which overload is chosen there is a lot of work that can be performed on any of the return types.

The setup is basically like func/func2 in this:

```
from typing import Literal, overload

@overload
def func(a: Literal[1]) -> str:
    ...

@overload
def func(a: Literal[2]) -> int:
    ...

def func(a: Literal[1, 2]) -> str | int:
    return "1" if a == 1 else 2

def func2(value: Literal[1, 2]) -> None:
    stuff = func(value)
    if value == 1:
        # Do specific stuff1
        reveal_type(stuff)
    else:
        # Do specific stuff2
         reveal_type(stuff)
    # Do general stuff

def func3(value: Literal[1, 2]) -> None:
    if value == 1:
        # Do specific stuff1 AND general stuff
        reveal_type(func(value))
    else:
        # Do specific stuff2 AND general stuff
         reveal_type(func(value))

func2(1)
func2(2)
func3(1)
func3(2)
```

func2 would allow me to reduce duplication, but does not type check because obviously the type checker does not remember that the narrowing of `stuff` depends on value. Meanwhile func3 typechecks but i have duplication.

Is there any chance that something like this could be introduced or am i missing (some obvious?) issues that this would introduce (besides the possible enormous work that might have to be done)?

[Playground link](https://pyright-play.net/?strict=true&locale=en&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2ANFGAG5mVjkAmAsAFBcACDTLDp1YlgUYAFcUAYwAU5AFyFiZKgG0AjAF0AlFAC0APigBnGCAVco1qADp7XXvxDM2XEWMkz5SoqQqUagBMugbGqDCWnDZ2DtzCouJScorK-uoatCF6RqbmUAA%2BmCiRVjYgJDASIChQAEQadZhi5FAAvG1QGlAklCYkUEGOCZ7JQbL0VBIkvioBmlmhuQByYCgzZdZmEsBinV5yk5TTOpvNUEfT7Z0aUTExAMRQACJgpggk0kjASNJ5O8ANGcYhVGFQAPrwD6yba7U7RGy9fp3e7WJ6vd6fb6-f67IYI1FQUEkCFQkgwqpws7ot5oEjrAK44DDDxJGQAZgmUxmaVUgUygyWxlW6xR1m%2BF251y6YvuNMxXx%2Bf1hgKgAEFls8oHSGVQmcDyiQwZRIYhyQcuccSDp4TEkRsCXKXm8TB9FTiVUF1ZrtfS%2BfrHfdiaSzbILZdrfCuAdxhp4THZEF48lOXHoynEzogA)

**Comment by erictraut** [comment]
I don't think this is feasible. It's unlikely that this would ever be implemented. See [this comment](https://github.com/microsoft/pyright/issues/5824#issuecomment-1694516536) for the reasoning.

The technique you're using in `func3` is the approach I recommend. If you have common code in both cases, perhaps you can try to refactor it into a subroutine to avoid duplication.
